src/net.cpp

```diff
@@ -923,16 +929,25 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
 
     // Protect the half of the remaining nodes which have been connected the longest.
     // This replicates the non-eviction implicit behavior, and precludes attacks that start later.
-    // Reserve half of these protected spots for localhost peers, even if
+    // Reserve half of these protected spots for onion and localhost peers, even if
     // they're not longest-uptime overall. This helps protect tor peers, which
     // tend to be otherwise disadvantaged under our eviction criteria.
-    size_t initial_size = vEvictionCandidates.size();
+    const size_t initial_size = vEvictionCandidates.size();
     size_t total_protect_size = initial_size / 2;
 
-    // Pick out up to 1/4 peers that are localhost, sorted by longest uptime.
-    std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), CompareLocalHostTimeConnected);
-    size_t local_erase_size = total_protect_size / 2;
-    vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.end() - local_erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return n.m_is_local; }), vEvictionCandidates.end());
+    // Pick out up to 1/4 peers that are connected via our tor onion service, sorted by longest uptime.
+    std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), CompareOnionTimeConnected);
+    const size_t local_erase_size = total_protect_size / 2;
+    vEvictionCandidates.erase(std::remove_if(vEvictionCandidates.end() - local_erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return n.m_is_onion; }), vEvictionCandidates.end());
+
+    // If no onion peers were removed, extend the same protection to localhost peers, as manually
+    // configured hidden services not using -bind will not be detected as inbound onion connections.
+    if (initial_size == vEvictionCandidates.size()) {
```

Comment:
Just to note - this assumes that if there are recognizable onion connections (via `-bind=addr:port=onion`) then there are **not** non-recognizable ones (statically/manually configured tor daemon pointing to our standard p2p port, e.g. 8333).

This assumption may not be true as there could be a node that is configured both via `-bind=addr:port=onion` (connections coming to 8334 are recognized as onion) and also a static tor config redirecting to 8333.

I think the code is ok as it is because the above would be rare and maybe not worth the code complexity to tweak the eviction policy for it.